python/lsst/daf/butler/_butler.py

```diff
@@ -878,17 +931,11 @@ def getUri(self, datasetRefOrType: Union[DatasetRef, DatasetType, str],
         TypeError
             Raised if no collections were provided.
         """
-        ref = self._findDatasetRef(datasetRefOrType, dataId, allowUnresolved=predict,
-                                   collections=collections, **kwds)
-        if ref.id is None:  # only possible if predict is True
-            if run is None:
-                run = self.run
-                if run is None:
-                    raise TypeError("Cannot predict location with run=None.")
-            # Lie about ID, because we can't guess it, and only
-            # Datastore.getUri() will ever see it (and it doesn't use it).
-            ref = ref.resolved(id=0, run=self.run)
-        return self.datastore.getUri(ref, predict)
+        primary, _ = self.getURIs(datasetRefOrType, dataId=dataId, predict=predict,
+                                  collections=collections, run=run, **kwds)
+        if primary is None:
+            raise RuntimeError(f"Found dataset but no single URI retrieved for it {datasetRefOrType}")
```

Comment:
I'm not sure whether this error message (which is repeated a few times in the code) means found dataset in registry, but either 0 or > 1 URI retrieved for it (related to previous question about what getURIs does if dataset in registry but 0 datastore entries), or whether found dataset in datastore (and registry), but found > 1 URI for it.

Reply:
This is specifically trapping for the case where someone is using getURI and the dataset does exist but it's disassembled so getURI is at that point useless to you. The `if None` here also convinces mypy that I'm never going to return `Optional[ButlerURI]` but always `ButlerURI`.